### ISSUE TITLE
Enhancement: export function to export libraries from single .mo files to file system representation of a given package

### ISSUE THREAD

**Issue by prfpeste** [OpenModelica/OpenModelica#12678]
Currently, when a library is created as a single .mo file, it cannot directly out of OM be converted into a library file system structure with package.mo and package.order etc. Here a converter or export function would be very useful. The vice versa may be also of use.

Thanks and regards
Peter

**Comment by casella** [comment]
@perost I guess the functionality to do this is already present in the interactive mode, we just need to make it accessible properly in OMEdit. Do I miss something?

**Comment by perost** [comment]
> @perost I guess the functionality to do this is already present in the interactive mode, we just need to make it accessible properly in OMEdit. Do I miss something?

I'm not aware of any such functionality in the compiler, OMEdit handles such things on its own I think.

**Comment by casella** [comment]
What I mean is, if you call the `save(className)` API function, the compiler saves the class to the file system. So, it must be aware of how the class is actually mapped onto it. How does this work?

**Comment by perost** [comment]
> What I mean is, if you call the `save(className)` API function, the compiler saves the class to the file system. So, it must be aware of how the class is actually mapped onto it. How does this work?

We know which file every class comes from of course,  and `save` just goes through the surrounding packages until it finds the outermost one with the same filename and saves that to the file.

So if we want a class to be saved in a different file we would just have to go through it and change all the `SourceInfo`:s to point to the new filename. But then you also have to take care to remove files that are no longer used to avoid getting conflicts.

I'm also unsure of what exactly is being asked for here, since there's no definitive structure for a library. It's usually a mix of files containing single classes and files containing whole packages. So automatically converting between single and multiple files might not be very useful, and we'd probably need something a bit more manual.

We could for example make it easier to change how a specific class should be saved, or even visualize the whole file structure and let the user make larger changes such as merging classes into the file above or moving specific classes to their own files. That seems like something that would be very nice to have, but would also require quite a lot of effort to implement well.

**Comment by casella** [comment]
Yeah, we need to discuss this with @adeas31. As I understand, we already have that kind of support for Duplicate, where OMEdit asks you this:
![immagine](https://github.com/OpenModelica/OpenModelica/assets/4233148/7859a1d2-f9fa-41a9-b468-5c329d5c2d97)

Unfortunately, you can't duplicate a package with the same name, so we currently can't use this feature to do what @prfpeste is asking for, but I guess it will be easy to add that provision to OMEdit.

**Comment by casella** [comment]
@prfpeste we can work on this in one-two months from now. For the time being, you can try the following:
- Open your MyPackage.mo
- Right-click on it in the Libraries browser, select Duplicate and give it another name, say MyPackage2, then select the structure you want for the file system mapping on the dialog shown above
- Save the duplicated package
- Rename the top directory name from MyPackage2 to MyPackage
- Use a smart editor (Notepad++, jEdit) to change all the instances of the string "MyPackage2" to "MyPackage" inside all the files of that directory

It's a bit ugly, but it should work.

**Comment by prfpeste** [comment]
@casella  great, thanks a lot, i.e. for the very fast reaction time and thank you for the description!
